src/day_3/solution.rs

```diff
@@ -0,0 +1,113 @@
+use std::collections::HashMap;
+
+fn find_item(pattern: &str, position: usize) -> String {
+  let mut item: String = "".to_string();
+  let mut found_item = false;
+  let mut repeat_trees = 0;
+
+  while !found_item {
+    let found_position = pattern.repeat(repeat_trees).char_indices().find(|(idx, _char)| *idx == position );
+
+    if found_position.is_none() {
+      repeat_trees += 1;
+
+      continue;
+    }
+
+    let (_idx, c) = found_position.unwrap();
+
+    item = c.to_string();
+    found_item = true;
+  }
+
+  return item;
+}
+
+// Ignore this, trying to reduce the amount of duplication with config maps
+// fn setup_map(position_to_add: usize, position: usize, line_to_add: usize) -> HashMap<String, usize> {
+//   let mut map = HashMap::new();
+
+//   map.insert("position_to_add".to_string(), position_to_add);
+//   map.insert("position".to_string(), position);
+//   map.insert("line_to_add".to_string(), line_to_add);
+
+//   return map;
+// }
+
+pub fn run() {
+  let mut reader = csv::Reader::from_path("src/day_3/input.csv").unwrap();
+
+  let tree_char = "#".to_string();
+  let mut position_1: usize = 0;
+  let mut position_3: usize = 0;
+  let mut position_5: usize = 0;
+  let mut position_7: usize = 0;
+  let mut position_2_down: usize = 0;
+  // Start at the value of the first line, which is 2
+  let mut position_2_down_line: u64 = 2;
+
+  let mut tree_tracker: HashMap<&str, usize> = HashMap::new();
+
+  for record in reader.records() {
+    let r = record.unwrap();
+    let slice_record = r.as_slice().to_owned();
+
+    let current_line = r.position().unwrap().line();
+
+    // Skip the first line
+    if current_line == 2 {
+      position_1 += 1;
+      position_3 += 3;
+      position_5 += 5;
+      position_7 += 7;
+      position_2_down += 1;
+      position_2_down_line += 2;
+
+      continue;
+    }
+
+    let item_1 = find_item(&slice_record, position_1);
+    let item_3 = find_item(&slice_record, position_3);
+    let item_5 = find_item(&slice_record, position_5);
+    let item_7 = find_item(&slice_record, position_7);
+
+
+    if item_1 == tree_char {
+      let count = tree_tracker.entry("position_1").or_insert(0);
```

Comment:
+1 for the `entry` api, pretty powerful

Reply:
:O 
I know no rust, but just looked this up - very cool feature of the language